examples/dft.masm

```diff
@@ -0,0 +1,288 @@
+    if.true
+        # n == 1
+        drop                                   # stack: []
+        push.0                                 # stack: [0]
+    else
+        dup                                    # stack: [n, n]
+        eq.2                                   # stack: [0 or 1, n]
+        if.true
+            # n == 2
+            drop                               # stack: []
+            push.1                             # stack: [1]
+        else
+            dup                                # stack: [n, n]
+            eq.4                               # stack: [0 or 1, n]
+            if.true
+                # n == 4
+                drop                           # stack: []
+                push.2                         # stack: [2]
+            else
+                dup                            # stack: [n, n]
+                eq.8                           # stack: [0 or 1, n]
+                if.true
+                    # n == 8
+                    drop                       # stack: []
+                    push.3                     # stack: [3]
+                else
+                    dup                        # stack: [n, n]
+                    eq.16                      # stack: [0 or 1, n]
+                    if.true
+                        # n == 16
+                        drop                       # stack: []
+                        push.4                     # stack: [4]
+                    else
+                        # n unsupported; halt
+                        push.1
+                        assertz
+                    end
+                end
+            end
+        end
+
+    end
+end
+
+#! Computes the primitive root of unity for a subgroup of size 2^n.
+#!
+#! Input: [log2_n], log2 of the size of the input list (must be <= 32)
+#! Output: [root_of_unity], where `root_of_unity` is the primitive root of unity of order 2^n
+#!
+#! Equivalent to Winterfell's implementation: 
+#! https://github.com/facebook/winterfell/blob/4543689f73a2efb9d30927535de7b4efe7e1802d/math/src/field/traits.rs#L254-L259
+proc.get_root_of_unity
+    push.TWO_ADIC_ROOT_OF_UNITY.TWO_ADICITY     # stack: [TWO_ADICITY, TWO_ADIC_ROOT_OF_UNITY, log2_n]
+    movup.2                                     # stack: [log2_n, TWO_ADICITY, TWO_ADIC_ROOT_OF_UNITY]
+    sub                                         # stack: [TWO_ADICITY - log2_n, TWO_ADIC_ROOT_OF_UNITY]
+    pow2                                        # stack: [2^(TWO_ADICITY - log2_n), TWO_ADIC_ROOT_OF_UNITY]
+    exp                                         # stack: [TWO_ADIC_ROOT_OF_UNITY^(2^(TWO_ADICITY - log2_n))]
+end
+
+#! Computes the kth element of the "frequency domain" (i.e. of the transformed list).
+#!
+#! Input: [k, root_of_unity, n, v_0, ..., v_{n-1}, ...]
+#! Output: [f_k, ...]
+#!
+#! Locals
+#! 0: k
+#! 1: root_of_unity
+#! 2: n (length of the list)
+#! 3: j (counter)
+#! 4: result (partial result of the computation)
+proc.f_k.5
```

Comment:
This is elegant. 

You could think of only storing `k, n, root_of_unity`. Only `j` and the `result` change with every iteration. But actually not sure if this would help. 

Nice implementation and nice comments!

Reply:
As discussed [here](https://github.com/0xPolygonMiden/examples/pull/150#discussion_r1397279064), if I were to redo, I would use stack operations instead of memory. But I don't think I should invest time in doing that at this point 🙂.

> Nice implementation and nice comments!

Thank you!